src/main/java/com/google/sps/servlets/CSVServlet.java

```diff
@@ -0,0 +1,70 @@
+package com.google.sps.servlets;
+ 
+import com.google.api.client.auth.oauth2.TokenResponseException;
+import com.google.appengine.api.datastore.PreparedQuery.TooManyResultsException;
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.ChromeOSDevice;
+import com.google.sps.gson.Json;
+import com.google.sps.servlets.Util;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.StringBuilder;
+import java.security.GeneralSecurityException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+ 
+@WebServlet("/csv")
+public class CSVServlet extends HttpServlet {
+ 
+  private UserService userService = UserServiceFactory.getUserService();
+  private Util utilObj = new Util();
+  public final String LOGIN_URL = "/login";
+  public final String HOME_URL = "/index.html";
+ 
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final User currentUser = userService.getCurrentUser();
+    if ((!userService.isUserLoggedIn()) || (currentUser == null)) {
+      response.sendRedirect(LOGIN_URL);
+      return;
+    }
+    final String userId = currentUser.getUserId();
+    response.setContentType("text/csv");
+    response.setHeader("Content-Disposition", "attachment; filename=\"my_devices.csv\"");
+    try {
+      OutputStream outputStream = response.getOutputStream();
+      final List<ChromeOSDevice> allDevices = utilObj.getAllDevices(userId);
+      final String outputResult = stringifyDevicesList(allDevices);
+      outputStream.write(outputResult.getBytes());
+      outputStream.flush();
+      outputStream.close();
+      response.setStatus(HttpServletResponse.SC_OK);
+    } catch(Exception e) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+    }
+  }
+
+  private String stringifyDevicesList(final List<ChromeOSDevice> allDevices) {
+    final String heading = "DEVICE ID,SERIAL NUMBER,ANNOTATED ASSET ID,ANNOTATED LOCATION,ANNOTATED USER\n";
+    final StringBuilder devicesStringBuilder = new StringBuilder(heading);
+    for (final ChromeOSDevice device : allDevices) {
+      devicesStringBuilder.append(device.getCSVLine());
+    }
+    return devicesStringBuilder.toString();
+  }
```

Comment:
The general approach here seems somewhat brittle to me, seeing as we must manually create the headings and we must manually ensure they are defined in the same order as in the `device.getCSVLine()` output.  One possibility is to use [java reflections](https://docs.oracle.com/javase/tutorial/reflect/member/index.html) to get a list of the objects members dynamically, which could be used to ensure all fields are correctly synchronized even if we add more fields to the ChromeOSDevice later on.  We can even get the fields' names via reflections which we could use filter which fields we export to CSV (e.g. a quick check of `if (csvFields.contains(field.getName()))` where `csvFields` is some set containing the fields we wish to export to CSV.

Reply:
While I agree that using reflections would make this more ready for change, at the end of the day I don't think this feature is going to change much at all. I have a hard time imagining us adding more fields to ChromeOSDevice that we would want to export within the lifetime of this project.